sql/grant_tables/grant_tables.go

```diff
@@ -67,6 +77,63 @@ func (g *GrantTables) GetTableNames(ctx *sql.Context) ([]string, error) {
 	return []string{"user"}, nil
 }
 
+// AuthMethod implements the interface mysql.AuthServer.
+func (g *GrantTables) AuthMethod(user string) (string, error) {
+	//TODO: this should pass in the host as well to correctly determine which auth method to use
+	return "mysql_native_password", nil
+}
+
+// Salt implements the interface mysql.AuthServer.
+func (g *GrantTables) Salt() ([]byte, error) {
+	return mysql.NewSalt()
+}
+
+// ValidateHash implements the interface mysql.AuthServer. This is called when the method used is "mysql_native_password".
+func (g *GrantTables) ValidateHash(salt []byte, user string, authResponse []byte, addr net.Addr) (mysql.Getter, error) {
+	if g.Disabled {
+		return mysqlGetter(user), nil
+	}
+
+	host, _, err := net.SplitHostPort(addr.String())
+	if err != nil {
+		return nil, err
+	}
+	//TODO: determine what the localhost is on the machine, then handle the conversion between ip and localhost
+	// For now, this just does another check for localhost if the host is 127.0.0.1
+	userRows := g.user.data.Get(UserPrimaryKey{
+		Host: host,
+		User: user,
+	})
+	if len(userRows) == 0 && host == "127.0.0.1" {
+		userRows = g.user.data.Get(UserPrimaryKey{
+			Host: "localhost",
+			User: user,
+		})
+	}
+	if len(userRows) != 1 {
+		return nil, mysql.NewSQLError(mysql.ERAccessDeniedError, mysql.SSAccessDeniedError, "Access denied for user '%v'", user)
+	}
+
+	if password, ok := userRows[0][40].(string); ok && len(password) > 0 { // index 40 is the authentication string, see the mysql.user schema
```

Comment:
I was thinking of refactoring this so that we’re returning objects rather than rows, which would remove the need for constants. I’ll have that in the next PR. I wasn’t too happy about this either.